plugins/inputs/vsphere/vsan.go

```diff
@@ -0,0 +1,539 @@
+	end := time.Now().UTC()
+
+	// We're using a fake metric key, since we only store one highwater mark per resource
+	start, ok := e.hwMarks.Get(hwMarksKeyPrefix+clusterRef.ref.Value, "generic")
+	if !ok {
+		// Look back 3 sampling periods by default
+		start = end.Add(time.Duration(e.Parent.MetricLookback) * time.Duration(-e.resourceKinds["vsan"].sampling) * time.Second)
+	}
+	e.Parent.Log.Debugf("[vSAN] Query vsan performance for time interval: %s ~ %s", start, end)
+	latest := start
+
+	for entityRefID := range metrics {
+		if !strings.HasPrefix(entityRefID, perfPrefix) {
+			continue
+		}
+		entityRefID = strings.TrimPrefix(entityRefID, perfPrefix)
+		var perfSpecs []vsantypes.VsanPerfQuerySpec
+
+		perfSpec := vsantypes.VsanPerfQuerySpec{
+			EntityRefId: fmt.Sprintf("%s:*", entityRefID),
+			StartTime:   &start,
+			EndTime:     &end,
+		}
+		perfSpecs = append(perfSpecs, perfSpec)
+
+		perfRequest := vsantypes.VsanPerfQueryPerf{
+			This:       perfManagerRef,
+			QuerySpecs: perfSpecs,
+			Cluster:    &clusterRef.ref,
+		}
+		resp, err := vsanmethods.VsanPerfQueryPerf(ctx, vsanClient, &perfRequest)
+
+		if err != nil {
+			if err.Error() == "ServerFaultCode: NotFound" {
+				e.Parent.Log.Errorf("[vSAN] Is vSAN performance service enabled for %s? Skipping ...", clusterRef.name)
+				break
+			}
+			e.Parent.Log.Errorf("[vSAN] Error querying performance data for %s: %s: %s.", clusterRef.name, entityRefID, err)
+			continue
+		}
+		tags := populateClusterTags(make(map[string]string), clusterRef, e.URL.Host)
+
+		count := 0
+		for _, em := range resp.Returnval {
+			vals := strings.Split(em.EntityRefId, ":")
+			var entityName, uuid string
+			if len(vals) == 1 {
+				entityName, uuid = vals[0], ""
+			} else {
+				entityName, uuid = vals[0], vals[1]
+			}
+
+			buckets := make(map[string]metricEntry)
+			tags := populateCMMDSTags(tags, entityName, uuid, cmmds)
+			var timeStamps []time.Time
+			// 1. Construct a timestamp list from sample info
+			formattedEntityName := hyphenReplacer.Replace(entityName)
+			for _, t := range strings.Split(em.SampleInfo, ",") {
+				tsParts := strings.Split(t, " ")
+				if len(tsParts) >= 2 {
+					// The return time string is in UTC time
+					utcTimeStamp := fmt.Sprintf("%sT%sZ", tsParts[0], tsParts[1])
+					ts, ok := time.Parse(time.RFC3339, utcTimeStamp)
+					if ok != nil {
+						e.Parent.Log.Errorf("[vSAN] Failed to parse a timestamp: %s. Skipping", utcTimeStamp)
+						timeStamps = append(timeStamps, time.Time{})
+						continue
+					}
+					timeStamps = append(timeStamps, ts)
+				}
+			}
+			// 2. Iterate on each measurement
+			for _, counter := range em.Value {
+				metricLabel := internal.SnakeCase(counter.MetricId.Label)
+				// 3. Iterate on each data point.
+				for i, values := range strings.Split(counter.Values, ",") {
+					ts := timeStamps[i]
+					if ts == (time.Time{}) {
+						continue
+					}
```

Comment:
This is very fragile. If you get your timestamps in "unexpected" format for at least _one_ timestamp, you might end-up with a `timeStamps` array _shorter_ than the `values` array and thus panic in line 296! Please be safe here and rather error out than rely on properties that are not _guaranteed_ by code.